### PR TITLE
ET-250 - Toggle preview mode.

### PR DESCRIPTION
**Pull request by makma**
Add support for toggling preview mode.

### CODE REVIEW

File: DancingGoat/Content/Styles/EditMode.less [hunk 1]
```diff
@@ -69,7 +66,11 @@
             }
         }
     }
-} 
+}
+
+.edin-in-cloud-button {
```
Comment (makma) [inline DancingGoat/Content/Styles/EditMode.less]: edin

Comment (daemoon) [reply DancingGoat/Content/Styles/EditMode.less]: also what exactly does edit in cloud mean?

File: DancingGoat/Areas/Admin/Controllers/SelfConfigController.cs [hunk 2]
```diff
@@ -29,13 +30,19 @@ public class SelfConfigController : Controller
         public const int PROJECT_EXISTENCE_VERIFICATION_REQUIRED_ITEMS = 32;
 
         protected readonly SelfConfigManager _selfConfigManager;
-        protected readonly IDeliveryClient client = new SampleDeliveryClient();
+        protected IDeliveryClient client { get; private set; }
```
Comment (PeterSkoda) [inline DancingGoat/Areas/Admin/Controllers/SelfConfigController.cs]: inconsistency in naming: do we prefix protected with underscore or not? (_selfConfigManager vs client)

File: DancingGoat/Controllers/ControllerBase.cs [hunk 3]
```diff
@@ -1,22 +1,29 @@
...
-        protected readonly IDeliveryClient baseClient = new SampleDeliveryClient();
-        protected readonly IDeliveryClient client;
+        protected IDeliveryClient baseClient { get; private set; }
+        protected IDeliveryClient client { get; private set; }
 
-        public ControllerBase()
+        protected override void Initialize(RequestContext requestContext)
         {
+            base.Initialize(requestContext);
+
```
Comment (daemoon) [inline DancingGoat/Controllers/ControllerBase.cs]: Emtpy lines.

File: DancingGoat/Scripts/PreviewMode/PreviewMode.js [hunk 4]
```diff
@@ -0,0 +1,107 @@
+﻿const previewModeToggleClassname = 'preview-mode-switch';
+const previewModeCookieName = 'IsPreviewMode';
```
Comment (daemoon) [inline DancingGoat/Scripts/PreviewMode/PreviewMode.js]: Inconsistent naming, what's the difference between preview and preview mode?

File: DancingGoat/Scripts/PreviewMode/PreviewMode.js [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+const previewApiKeyCookieName = 'PreviewApiKey';
+const enterApiKeyFirstMessage = 'Enter your Preview API key';
+const enterApiKeyNextMessage = 'Enter your Delivery Preview API key -- you can find it in Kentico Cloud under Project settings -> API keys';
```
Comment (daemoon) [inline DancingGoat/Scripts/PreviewMode/PreviewMode.js]: Maybe rather Kentico Cloud App?

File: DancingGoat/Scripts/PreviewMode/PreviewMode.js [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+const enterApiKeyPromptTitle = 'Preview API key';
+
+const previewModeToggles = document.getElementsByClassName(previewModeToggleClassname);
+const isPreviewModeEnabled = getCookie(previewModeCookieName);
+for (toggle of previewModeToggles) {
```
Comment (daemoon) [inline DancingGoat/Scripts/PreviewMode/PreviewMode.js]: What the heck? Why are you iterating here?

File: DancingGoat/Controllers/ControllerBase.cs [hunk 7]
```diff
@@ -51,6 +58,13 @@ public static DeliveryClient CreateDeliveryClient()
             var options = provider.GetDeliveryOptions();
 
             options.ProjectId = ProjectUtils.GetProjectId();
+
+            if (ProjectUtils.IsInPreviewMode(httpContext) && !string.IsNullOrEmpty(ProjectUtils.GetPreviewApiKey(httpContext)))
+            {
+                options.UsePreviewApi = true;
+                options.PreviewApiKey = ProjectUtils.GetPreviewApiKey(httpContext);
```
Comment (michalpaukert) [inline DancingGoat/Controllers/ControllerBase.cs]: You are getting PreviewApiKey twice

File: DancingGoat/Utils/ProjectUtils.cs [hunk 8]
```diff
@@ -1,11 +1,24 @@
...
 {
     public class ProjectUtils
     {
+        private static string AppSettingPreviewApiKey = "PreviewApiKey";
+        private static string PreviewApiKeyCookieName = "PreviewApiKey";
+
         public static string GetProjectId()
         {
+            // check web.config first
+            if (Guid.TryParse(ConfigurationManager.AppSettings["ProjectId"], out Guid projectId))
```
Comment (michalpaukert) [inline DancingGoat/Utils/ProjectUtils.cs]: I think this block of code should be only in master.

File: DancingGoat/Utils/ProjectUtils.cs [hunk 9]
```diff
@@ -16,5 +29,51 @@ public static string GetProjectId()
 
             return decodedGuid.ToString();
         }
+
+        public static string GetPreviewApiKey(HttpContext httpContext)
+        {
+            // check web.config first
+            if (ConfigurationManager.AppSettings.AllKeys.Contains(AppSettingPreviewApiKey))
```
Comment (michalpaukert) [inline DancingGoat/Utils/ProjectUtils.cs]: I think this block of code should be only / even in master.

File: DancingGoat/Utils/ProjectUtils.cs [hunk 10]
```diff
@@ -16,5 +29,51 @@ public static string GetProjectId()
...
+        public static string GetPreviewApiKey(HttpContext httpContext)
+        {
+            // check web.config first
+            if (ConfigurationManager.AppSettings.AllKeys.Contains(AppSettingPreviewApiKey))
+            {
+                var previewApiKey = ConfigurationManager.AppSettings[AppSettingPreviewApiKey];
+
+                return previewApiKey;
+            }
+
```
Comment (michalpaukert) [inline DancingGoat/Utils/ProjectUtils.cs]: empty line

File: DancingGoat/Views/Shared/_Layout.cshtml [hunk 11]
```diff
@@ -13,9 +11,10 @@
 <head id="head">
     <!-- Google Tag Manager -->
     <script>
-        (function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start': new Date().getTime(), event: 'gtm.js'});
-        var f = d.getElementsByTagName(s)[0], j = d.createElement(s), dl = l != 'dataLayer' ? '&l=' + l : '';
-        j.async = true; j.src = 'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
+        (function (w, d, s, l, i) {
```
Comment (michalpaukert) [inline DancingGoat/Views/Shared/_Layout.cshtml]: this code block may not be changed

File: DancingGoat/Views/Shared/EditModePanel.cshtml [hunk 12]
```diff
@@ -1,18 +1,31 @@
...
+        <div class="edit-toggle__caption">
+            <span class="edit-toggle__title">Edit mode</span>
         </div>
-        <div class="edit-mode-panel__caption">
-            <span class="edit-mode-panel__title">Edit mode</span>
+    </div>
+    <div class="preview-toggle">
+        <label class="preview-toggle__switch">
+            <input class="preview-mode-switch" type="checkbox" onchange="togglePreviewMode(this.checked, '@ProjectGuid', true)
+                   " autocomplete="off">
```
Comment (michalpaukert) [inline DancingGoat/Views/Shared/EditModePanel.cshtml]: " autocomplete="off"

File: DancingGoat/Utils/ProjectUtils.cs [hunk 13]
```diff
@@ -16,5 +29,51 @@ public static string GetProjectId()
...
+            // check web.config first
+            bool.TryParse(ConfigurationManager.AppSettings["UsePreviewApi"], out var isPreviewFromConfig);
+
+            if (isPreviewFromConfig) return true;
+
+            // check preview mode from cookie
+            var isPreviewModeEnabledCookieName = "IsPreviewMode";
+            var cookies = httpContext.Request.Cookies;
+
+            if (cookies.AllKeys.Contains(isPreviewModeEnabledCookieName) && !String.IsNullOrEmpty(cookies[isPreviewModeEnabledCookieName].Value))
```
Comment (michalpaukert) [inline DancingGoat/Utils/ProjectUtils.cs]: sometimes you are using string.Empty and sometimes String.Empty

File: DancingGoat/Scripts/PreviewMode/PreviewMode.js [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+    const message = isFirstEnterOfApiKey ? enterApiKeyFirstMessage : enterApiKeyNextMessage;
+    const apiKey = prompt(message, enterApiKeyPromptTitle);
+    if (apiKey === null) {
+        const previewModeToggles = document.getElementsByClassName(previewModeToggleClassname);
+        for (toggle of previewModeToggles) {
+            toggle.checked = false
+        }
+    }
+
+    if (apiKey !== null) {
```
Comment (daemoon) [inline DancingGoat/Scripts/PreviewMode/PreviewMode.js]: At this point we can be pretty certain that apiKey is not null, see line 33

File: DancingGoat/Areas/Admin/Controllers/SelfConfigController.cs (hunk 2 above)
Comment (makma) [reply DancingGoat/Areas/Admin/Controllers/SelfConfigController.cs]: IDK - eg.: 'protected baseClient' without underscore.

File: DancingGoat/Scripts/PreviewMode/PreviewMode.js [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+function validateApiKey(apiKey, projectId) {
+    const deliverRequestUrl = 'https://preview-deliver.kenticocloud.com/' + projectId + '/items/home';
+    const xhr = new XMLHttpRequest();
+    xhr.open('GET', deliverRequestUrl, true);
+    xhr.setRequestHeader("authorization", `Bearer ${apiKey}`);
+    xhr.onreadystatechange = function () {
+        if (xhr.readyState === 4 && xhr.status === 200) {
+            logEnterPreviewKeyResult(true);
```
Comment (daemoon) [inline DancingGoat/Scripts/PreviewMode/PreviewMode.js]: Too much responsibility for this function which only does validate API key

File: DancingGoat/Scripts/PreviewMode/PreviewMode.js [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+
+function setPreviewApiKeyCookie(apiKey) {
+    document.cookie = `PreviewApiKey=${apiKey}; max-age=31536000; path=/`;
+}
+
+function logAttemptToEnterPreviewKey(isFirstAttempt) {
+    const label = isFirstAttempt ? 'first' : 'another';
+    ga('send', 'event', {
+        eventCategory: 'Administrative section',
+        eventAction: 'Enter Api key attempt',
```
Comment (daemoon) [inline DancingGoat/Scripts/PreviewMode/PreviewMode.js]: Preview API key

File: DancingGoat/Scripts/PreviewMode/PreviewMode.js [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+        eventLabel: label,
+    });
+}
+
+function logEnterPreviewKeyResult(isSuccessful) {
+    const result = isSuccessful ? 'succesful' : 'not-succesful' 
+    ga('send', 'event', {
+        eventCategory: 'Administrative section',
+        eventAction: 'Enter Api key result',
```
Comment (daemoon) [inline DancingGoat/Scripts/PreviewMode/PreviewMode.js]: Preview.

File: DancingGoat/Scripts/PreviewMode/PreviewMode.js [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+function enterApiKey(isFirstEnterOfApiKey, projectId) {
+    logAttemptToEnterPreviewKey(isFirstEnterOfApiKey);
+    const message = isFirstEnterOfApiKey ? enterApiKeyFirstMessage : enterApiKeyNextMessage;
+    const apiKey = prompt(message, enterApiKeyPromptTitle);
+    if (apiKey === null) {
+        const previewModeToggles = document.getElementsByClassName(previewModeToggleClassname);
+        for (toggle of previewModeToggles) {
+            toggle.checked = false
```
Comment (michalpaukert) [inline DancingGoat/Scripts/PreviewMode/PreviewMode.js]: missing ;

File: DancingGoat/Utils/ProjectUtils.cs (hunk 8 above)
Comment (daemoon) [inline DancingGoat/Utils/ProjectUtils.cs]: Why was this necessary?

File: DancingGoat/Content/Styles/EditMode.less (hunk 1 above)
Comment (makma) [reply DancingGoat/Content/Styles/EditMode.less]: I'm not sure I understand - it's button navigating to listing of current project.
```
> <div class="edin-in-cloud-button">
>   <a target="_blank" href="https://app.kenticocloud.com/ddd5e846-5685-0067-5338-c56954c0bd7c" class="btn-cloud btn-cloud__padding btn-cloud-info">
>     <i class="btn-pen"></i> Edit in Kentico Cloud
>   </a>
> </div>
```

File: DancingGoat/Controllers/ControllerBase.cs (hunk 3 above)
Comment (makma) [reply DancingGoat/Controllers/ControllerBase.cs]: TODO: fix

File: DancingGoat/Views/Shared/EditModePanel.cshtml (hunk 12 above)
Comment (daemoon) [inline DancingGoat/Views/Shared/EditModePanel.cshtml]: Seems bit off.

File: DancingGoat/Scripts/PreviewMode/PreviewMode.js [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+const enterApiKeyNextMessage = 'Enter your Delivery Preview API key -- you can find it in Kentico Cloud under Project settings -> API keys';
+const enterApiKeyPromptTitle = 'Preview API key';
+
+const previewModeToggles = document.getElementsByClassName(previewModeToggleClassname);
+const isPreviewModeEnabled = getCookie(previewModeCookieName);
+for (toggle of previewModeToggles) {
+    toggle.checked = isPreviewModeEnabled === 'true' ? true : false;
+}
+
+function togglePreviewMode(isChecked, projectId, isFirstEnterOfApiKey) {
```
Comment (PeterSkoda) [inline DancingGoat/Scripts/PreviewMode/PreviewMode.js]: I believe that this does not really togglePreviewMode, rather than "tries" to do it

File: DancingGoat/Views/Shared/EditModePanel.cshtml [hunk 20]
```diff
@@ -1,18 +1,31 @@
...
+            <span class="slider round"></span>
+        </label>
+        <div class="preview-toggle__caption">
+            <span class="preview-toggle__title">Preview mode</span>
         </div>
     </div>
-    <div class="edit-mode-panel__navigate-link">
+
+    <div class="footer-navigate-link">
```
Comment (daemoon) [inline DancingGoat/Views/Shared/EditModePanel.cshtml]: Used to be Element, now it is incosistent, as it is footer in edit mode panel

File: DancingGoat/Scripts/PreviewMode/PreviewMode.js (hunk 19 above)
Comment (PeterSkoda) [inline DancingGoat/Scripts/PreviewMode/PreviewMode.js]: maybe isFirstEnterOfApiKey rename to isFirstTry ?

File: DancingGoat/Controllers/ControllerBase.cs (hunk 7 above)
Comment (makma) [reply DancingGoat/Controllers/ControllerBase.cs]: It's just consistency check, it's synchronized imo OK.

File: DancingGoat/Areas/Admin/Controllers/SelfConfigController.cs (hunk 2 above)
Comment (makma) [reply DancingGoat/Areas/Admin/Controllers/SelfConfigController.cs]: Fixed

File: DancingGoat/Content/Styles/EditMode.less (hunk 1 above)
Comment (makma) [reply DancingGoat/Content/Styles/EditMode.less]: Fixed

File: DancingGoat/Controllers/ControllerBase.cs (hunk 3 above)
Comment (makma) [reply DancingGoat/Controllers/ControllerBase.cs]: Fixed

File: DancingGoat/Scripts/PreviewMode/PreviewMode.js (hunk 5 above)
Comment (makma) [reply DancingGoat/Scripts/PreviewMode/PreviewMode.js]: This copy is provided by Aaron, I'll keep it as is.

File: DancingGoat/Scripts/PreviewMode/PreviewMode.js (hunk 19 above)
Comment (makma) [reply DancingGoat/Scripts/PreviewMode/PreviewMode.js]: Renamed to tryTogglePreviewMode.

Comment (makma) [reply DancingGoat/Scripts/PreviewMode/PreviewMode.js]: Fixed

File: DancingGoat/Scripts/PreviewMode/PreviewMode.js (hunk 18 above)
Comment (makma) [reply DancingGoat/Scripts/PreviewMode/PreviewMode.js]: Fixed

File: DancingGoat/Scripts/PreviewMode/PreviewMode.js [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+    }
+}
+
+function enterApiKey(isFirstEnterOfApiKey, projectId) {
+    logAttemptToEnterPreviewKey(isFirstEnterOfApiKey);
+    const message = isFirstEnterOfApiKey ? enterApiKeyFirstMessage : enterApiKeyNextMessage;
+    const apiKey = prompt(message, enterApiKeyPromptTitle);
+    if (apiKey === null) {
+        const previewModeToggles = document.getElementsByClassName(previewModeToggleClassname);
+        for (toggle of previewModeToggles) {
```
[the remark this answers is not in the document]
Comment (makma) [reply DancingGoat/Scripts/PreviewMode/PreviewMode.js]: Fixed

File: DancingGoat/Scripts/PreviewMode/PreviewMode.js (hunk 6 above)
Comment (makma) [reply DancingGoat/Scripts/PreviewMode/PreviewMode.js]: Fixed

File: DancingGoat/Scripts/PreviewMode/PreviewMode.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (makma) [reply DancingGoat/Scripts/PreviewMode/PreviewMode.js]: Next time pls try to suggest better name if you are not satisfied with the current one. thx

Renamed to isNewStateEnabled.

File: DancingGoat/Scripts/PreviewMode/PreviewMode.js [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+for (toggle of previewModeToggles) {
+    toggle.checked = isPreviewModeEnabled === 'true' ? true : false;
+}
+
+function togglePreviewMode(isChecked, projectId, isFirstEnterOfApiKey) {
+    if (isChecked) {
+        const apiKey = getCookie(previewApiKeyCookieName);
+        if (apiKey === '') {
+            enterApiKey(isFirstEnterOfApiKey, projectId);
+        } else {
```
[the remark this answers is not in the document]
Comment (makma) [reply DancingGoat/Scripts/PreviewMode/PreviewMode.js]: I know but I want the else block preserve - it could help to recover from inconsistent cookie state (weird extension, user manipulation...)

File: DancingGoat/Scripts/PreviewMode/PreviewMode.js [hunk 23]
```diff
@@ -0,0 +1,107 @@
...
+        if (xhr.readyState === 4 && xhr.status === 401) {
+            logEnterPreviewKeyResult(false);
+            togglePreviewMode(true, projectId, false);
+        }
+    };
+    xhr.send();
+}
+
+
+function getCookie(cname) {
```
[the remark this answers is not in the document]
Comment (makma) [reply DancingGoat/Scripts/PreviewMode/PreviewMode.js]: Yes it's not ;). At first I wanted to use jquery's $.cookie() but it's not possible since cookie module must be added. So I decided to go this way.

ref.: https://www.w3schools.com/js/js_cookies.asp

File: DancingGoat/Scripts/PreviewMode/PreviewMode.js (hunk 16 above)
Comment (makma) [reply DancingGoat/Scripts/PreviewMode/PreviewMode.js]: Fixed

File: DancingGoat/Scripts/PreviewMode/PreviewMode.js (hunk 17 above)
Comment (makma) [reply DancingGoat/Scripts/PreviewMode/PreviewMode.js]: Fixed

File: DancingGoat/Utils/ProjectUtils.cs (hunk 8 above)
Comment (makma) [reply DancingGoat/Utils/ProjectUtils.cs]: I think so as well, we should consider it after closing this task.

Comment (makma) [reply DancingGoat/Utils/ProjectUtils.cs]: To ensure consistent behavior (see isPreview flag in if statement when rendering Edit mode panel).

File: DancingGoat/Views/Shared/_Layout.cshtml (hunk 11 above)
Comment (makma) [reply DancingGoat/Views/Shared/_Layout.cshtml]: It's just 'prettified' by VS, it works and does the same thing as before. But I've reverted it to original gtm script.

File: DancingGoat/Scripts/PreviewMode/PreviewMode.js (hunk 4 above)
Comment (makma) [reply DancingGoat/Scripts/PreviewMode/PreviewMode.js]: I don't get it. Could you more elaborate on this issue?

File: DancingGoat/Scripts/PreviewMode/PreviewMode.js (hunk 14 above)
Comment (makma) [reply DancingGoat/Scripts/PreviewMode/PreviewMode.js]: Do you try to suggest it should be in else block?

File: DancingGoat/Areas/Admin/Controllers/SelfConfigController.cs (hunk 2 above)
Comment (makma) [reply DancingGoat/Areas/Admin/Controllers/SelfConfigController.cs]: Fixed to `SelfConfigManager `and `Client`.

File: DancingGoat/Scripts/PreviewMode/PreviewMode.js [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+        location.reload();
+    }
+}
+
+function enterApiKey(isFirstEnterOfApiKey, projectId) {
+    logAttemptToEnterPreviewKey(isFirstEnterOfApiKey);
+    const message = isFirstEnterOfApiKey ? enterApiKeyFirstMessage : enterApiKeyNextMessage;
+    const apiKey = prompt(message, enterApiKeyPromptTitle);
+    if (apiKey === null) {
+        const previewModeToggles = document.getElementsByClassName(previewModeToggleClassname);
```
[the remark this answers is not in the document]
Comment (makma) [reply DancingGoat/Scripts/PreviewMode/PreviewMode.js]: Fixed, removed `for` iteration as there is only one preview toggle at a time.

File: DancingGoat/Utils/ProjectUtils.cs [hunk 25]
```diff
@@ -16,5 +29,51 @@ public static string GetProjectId()
...
+            {
+                var isInPreviewCookieValue = cookies.Get(isPreviewModeEnabledCookieName).Value;
+                if (isInPreviewCookieValue == "true")
```
[the remark this answers is not in the document]
Comment (makma) [reply DancingGoat/Utils/ProjectUtils.cs]: I'll like it more this way - imo it's more obvious and more readable. On prod it'll be minified so it doesn't matter it's longer.

File: DancingGoat/Utils/ProjectUtils.cs [hunk 26]
```diff
@@ -16,5 +29,51 @@ public static string GetProjectId()
 
             return decodedGuid.ToString();
         }
+
+        public static string GetPreviewApiKey(HttpContext httpContext)
+        {
+            // check web.config first
+            if (ConfigurationManager.AppSettings.AllKeys.Contains(AppSettingPreviewApiKey))
+            {
+                var previewApiKey = ConfigurationManager.AppSettings[AppSettingPreviewApiKey];
```
[the remark this answers is not in the document]
Comment (makma) [reply DancingGoat/Utils/ProjectUtils.cs]: 🐱‍🐉

File: DancingGoat/Views/Shared/EditModePanel.cshtml (hunk 20 above)
Comment (makma) [reply DancingGoat/Views/Shared/EditModePanel.cshtml]: You might be right, renamed to:
`edit-mode-panel__navigate-link`

File: DancingGoat/Utils/ProjectUtils.cs (hunk 25 above)
[the remark this answers is not in the document]
Comment (PeterSkoda) [reply DancingGoat/Utils/ProjectUtils.cs]: I still think one-liner return isInPreviewCookieValue == "true" would be better
